plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/functions/tablechanges/TableChangesSplitSource.java

```diff
@@ -62,71 +62,102 @@ public TableChangesSplitSource(
     {
         tableLocation = functionHandle.tableLocation();
         fileSystem = fileSystemFactory.create(session);
-        currentVersion = new AtomicLong(functionHandle.firstReadVersion());
         tableReadVersion = functionHandle.tableReadVersion();
         transactionLogDir = getTransactionLogDir(tableLocation);
+        currentVersion = new AtomicLong(extractCurrentVersion(functionHandle));
+    }
+
+    private long extractCurrentVersion(TableChangesTableFunctionHandle functionHandle)
+    {
+        if (functionHandle.firstReadVersion().isPresent()) {
+            return functionHandle.firstReadVersion().get();
+        }
+        long currentVersion = 0;
+        while (currentVersion <= functionHandle.tableReadVersion()) {
```

Comment:
This should walk backwards from the present, not forwards from the beginning of the table history.
It should be possible to only have a week of table history but still read those CDC records.